packages/doenetml-worker-rust/lib-doenetml-core/src/core/component_builder.rs

```diff
@@ -362,7 +376,60 @@ impl ComponentBuilder {
                         unrecognized_attributes: HashMap::new(),
                     },
                 );
-                if let Some(Source::Ref(ref_resolution)) = &elm.extending {
+
+                // A some of a component's attributes may specify that refs in those attributes should not be expanded.
+                // We mark the corresponding components.
+                // Because the component's attributes have not been processed yet, we need to look up the corresponding
+                // attributes from `node`
+                for preserve_ref_attr_name in component
+                    .get_preserve_ref_attribute_indices()
+                    .iter()
+                    .map(|idx| component.get_attribute_names()[*idx])
+                {
+                    let attr = elm
+                        .attributes
+                        .iter()
+                        .find(|a| a.name == preserve_ref_attr_name);
+                    if attr.is_none() {
+                        continue;
+                    }
+                    let attr = attr.unwrap();
+                    for child in &attr.children {
+                        match child {
+                            UntaggedContent::Ref(idx) => {
+                                self.component_preserve_refs[ComponentIdx::from(*idx)] = true;
+                            }
+                            _ => {}
+                        }
+                    }
+                }
+
+                if self.component_preserve_refs[component.get_idx()] {
```

Comment:
The reason we know we will have set `component_preserve_refs` before we get to `component` is because we are processing components starting with the lowest index, which means we'll process a component before we process the components inside its attributes. We should include a comment here to that effect.

More importantly, though, is to add a comment before line 324 indicating the requirement that we attempt to create nodes lowest-index first. The current comment before line 324 would lead one to think it is optional (as it was when that comment was created).

I originally thought that we'd run into trouble with self references, such as `<xref name="x" ref="$x" />`, as in this case the attribute will get created before the `<xref>`. However, the critical code from just above will get run the first time through trying to make `<xref>` even though that try ends up with an `Err()`. I recommend some comment indicating that we need to make sure the above does run even if creating `<xref>` fails. More importantly, we should have a test checking that it works even with such a self-reference.

Reply:
Not exactly sure what the last paragraph is asking me to do, but I added a test of a self-referential xref.